File: 0090-subsets-ii/0090-subsets-ii.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsetsWithDup = function(nums) {
    nums.sort((a,b) => b-a)
    let ans = []

    var solve = function(i, ds){
        ans.push([...ds])

        for(let j=i ; j<nums.length ; j++){
            if(j>i && nums[j] === nums[j-1]) continue
        
            ds.push(nums[j])
            solve(j+1, ds)
            ds.pop()
        }
    }

    solve(0, [])

    return ans
};

File: 0090-subsets-ii/0090-subsets-ii.cpp
class Solution {
public:
    void solve(int i, vector<int>& ds, vector<int>& nums, vector<vector<int>>& ans, set<vector<int>>& stt){
        if(i==nums.size()){
            if(stt.find(ds) != stt.end()){
                return;
            }

            ans.push_back(ds);
            stt.insert(ds);
            return;
        }

        ds.push_back(nums[i]);
        solve(i+1, ds, nums, ans, stt);
        ds.pop_back();

        solve(i+1, ds, nums, ans, stt);
    }
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());

        vector<vector<int>> ans;
        vector<int> ds;
        set<vector<int>> stt;

        solve(0, ds, nums, ans, stt);

        return ans;
    }
};